sdk/pinocchio/src/sysvars/slot_hashes/mod.rs

```diff
@@ -0,0 +1,443 @@
+//! Efficient, zero-copy access to SlotHashes sysvar data.
+
+#[cfg(test)]
+mod tests;
+
+use crate::{
+    account_info::{AccountInfo, Ref},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+    sysvars::clock::Slot,
+};
+use core::{mem, ops::Deref, slice::from_raw_parts};
+#[cfg(feature = "std")]
+use std::boxed::Box;
+
+/// SysvarS1otHashes111111111111111111111111111
+pub const SLOTHASHES_ID: Pubkey = [
+    6, 167, 213, 23, 25, 47, 10, 175, 198, 242, 101, 227, 251, 119, 204, 122, 218, 130, 197, 41,
+    208, 190, 59, 19, 110, 45, 0, 85, 32, 0, 0, 0,
+];
+/// Number of bytes in a hash.
+pub const HASH_BYTES: usize = 32;
+/// Sysvar data is:
+/// len    (8 bytes): little-endian entry count (≤ 512)
+/// entries(len × 40 bytes):    consecutive `(u64 slot, [u8;32] hash)` pairs
+/// Size of the entry count field at the beginning of sysvar data.
+pub const NUM_ENTRIES_SIZE: usize = mem::size_of::<u64>();
+/// Size of a slot number in bytes.
+pub const SLOT_SIZE: usize = mem::size_of::<Slot>();
+/// Size of a single slot hash entry (slot + hash).
+pub const ENTRY_SIZE: usize = SLOT_SIZE + HASH_BYTES;
+/// Maximum number of slot hash entries that can be stored in the sysvar.
+pub const MAX_ENTRIES: usize = 512;
+/// Max size of the sysvar data in bytes. 20488. Golden on mainnet (never smaller)
+pub const MAX_SIZE: usize = NUM_ENTRIES_SIZE + MAX_ENTRIES * ENTRY_SIZE;
+
+/// A single entry in the `SlotHashes` sysvar.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[repr(C)]
+pub struct SlotHashEntry {
+    /// The slot number stored as little-endian bytes.
+    slot_le: [u8; 8],
+    /// The hash corresponding to the slot.
+    pub hash: [u8; HASH_BYTES],
+}
+
+const _: [(); 1] = [(); mem::align_of::<SlotHashEntry>()];
+
+/// SlotHashes provides read-only, zero-copy access to SlotHashes sysvar bytes.
+pub struct SlotHashes<T: Deref<Target = [u8]>> {
+    data: T,
+    /// Number of entries (decoded from the 8-byte prefix).  Immutable.
+    len: usize,
+}
+
+/// Reads the entry count from the first 8 bytes of data.
+/// Returns None if the data is too short.
+#[inline(always)]
+pub(crate) fn read_entry_count_from_bytes(data: &[u8]) -> Option<usize> {
+    if data.len() < NUM_ENTRIES_SIZE {
+        return None;
+    }
+    Some(unsafe { u64::from_le_bytes(*(data.as_ptr() as *const [u8; 8])) } as usize)
```

Comment:
*minor*

As the size check above uses `NUM_ENTRIES_SIZE`, it seems reasonable to use the same constant when reading the value as well.

```suggestion
    Some(unsafe { u64::from_le_bytes(*(data.as_ptr() as *const [u8; NUM_ENTRIES_SIZE])) } as usize)
```

Reply:
d422291c5af5263155b5412b663c347478ab3258